src/neptune_api/client.py

```diff
@@ -394,3 +399,103 @@ def sync_auth_flow(self, request: httpx.Request) -> Generator[httpx.Request, htt
     async def async_auth_flow(self, request: httpx.Request) -> typing.AsyncGenerator[httpx.Request, httpx.Response]:
         # TODO: Missing implementation
         yield request
+
+
+USE_IAP = os.environ.get("NEPTUNE_GCP_USE_IAP", False) in {"true", "t", "1", "True"}
+
+if USE_IAP:
+    import google.auth
+    from google.cloud import iam_credentials_v1
+
+    NEPTUNE_IAP_SERVICE_ACCOUNT = os.environ["NEPTUNE_IAP_SERVICE_ACCOUNT"]
+
+    class IdentityAwareProxyAuthenticator(httpx.Auth):
+        def __init__(
+            self,
+            service_account_email: str,
+            additional_authenticator: Optional[httpx.Auth] = None,
+        ):
+            self._additional_authenticator = additional_authenticator
+            self._service_account_email = service_account_email
+            self._tokens: Dict[str, str] = {}
+
+        def _generate_jwt_payload(self, resource_url: str) -> str:
+            iat = datetime.datetime.now(tz=datetime.timezone.utc)
+            exp = iat + timedelta(hours=1)
+            return json.dumps(
+                {
+                    "iss": self._service_account_email,
+                    "sub": self._service_account_email,
+                    "aud": resource_url,
+                    "iat": int(iat.timestamp()),
+                    "exp": int(exp.timestamp()),
+                }
+            )
+
+        def _sign_jwt(self, resource_url: str) -> str:
+            source_credentials, _ = google.auth.default()
+            iam_client = iam_credentials_v1.IAMCredentialsClient(credentials=source_credentials)
+            payload = self._generate_jwt_payload(resource_url)
+            jwt: str = iam_client.sign_jwt(
+                name=iam_client.service_account_path("-", self._service_account_email),
+                payload=payload,
+            ).signed_jwt
+            return jwt
+
+        def sync_auth_flow(self, request: httpx.Request) -> Generator[httpx.Request, httpx.Response, None]:
+            url = str(request.url.copy_with(params=None))
+            token = self._tokens.get(url)
+            if token is None:
+                token = self._sign_jwt(url)
```

Comment:
Am I reading this correctly: we fetch a new JWT with each request?

Reply:
for each url